Make DispatchKeyGuards accept DispatchKey::Undefined

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32788 Add XLAPreAutograd key for XLA use cases that need custom autograd.
* #32787 Stop using dispatchTypeId to do checks for tensor list unwrap.
* #32734 Centralize addition of "always on" dispatch keys.
* **#32729 Make DispatchKeyGuards accept DispatchKey::Undefined**
* #32728 Rename DispatchKey::UndefinedTensorId to Undefined

When working on the vmap prototype I noticed that this was helpful
as it lets me easily initialize a no-op guard, if I need to do it
at constructor time (which I usually do, because the guards don't
have move constructors).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628092](https://our.internmc.facebook.com/intern/diff/D19628092)

ezyang committed:
Make DispatchKeyGuards accept DispatchKey::Undefined

When working on the vmap prototype I noticed that this was helpful
as it lets me easily initialize a no-op guard, if I need to do it
at constructor time (which I usually do, because the guards don't
have move constructors).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

smessmer reviewed (approved):
I'm ok with landing this if you want to, but I think it unnecessarily complicates the concept.

An alternative for your use case might be to have a local helper function in your test that returns a c10::optional<IncludeDispachKeyGuard> and is either set or not set.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 7c96e788:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7c96e7881482b127c272f7751b36917e973601d9).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4363072) pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build (1/1)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92191053) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92191053)) :snowflake:
<details>
<summary>
<code>E: Failed to fetch https://nvidia.github.io/libnvidia-container/ubuntu16.04/amd64/./libnvidia-container1_1.0.7-1_amd64.deb  gnutls_handshake() failed: Error in the pull function.</code>
</summary>

```
89% [25 linux-modules-extra-4.4.0-173-generic 32.2 MB/36.6 MB 88%] 92% [Working]                                                      20.6 MB/s 0s                                                                                 Get:26 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 intel-microcode amd64 3.20191115.1ubuntu0.16.04.2 [2,408 kB] 
92% [26 intel-microcode 24.6 kB/2,408 kB 1%]                       20.6 MB/s 0s 95% [Working]                                                      20.6 MB/s 0s                                                                                 Get:27 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 amd64-microcode amd64 3.20191021.1+really3.20180524.1~ubuntu0.16.04.2 [30.8 kB] 
95% [27 amd64-microcode 23.5 kB/30.8 kB 76%]                       20.6 MB/s 0s 95% [Working]                                                      20.6 MB/s 0s                                                                                 Get:28 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 linux-image-generic amd64 4.4.0.173.181 [2,486 B] 
95% [28 linux-image-generic 2,297 B/2,486 B 92%]                   20.6 MB/s 0s 96% [Working]                                                      20.6 MB/s 0s                                                                                 Get:29 http://archive.ubuntu.com/ubuntu xenial/universe amd64 moreutils amd64 0.57-1 [55.0 kB] 
96% [29 moreutils 2,295 B/55.0 kB 4%]                              20.6 MB/s 0s 96% [Working]                                                      20.6 MB/s 0s                                                                                 Get:30 http://archive.ubuntu.com/ubuntu xenial/main amd64 tcl8.6 amd64 8.6.5+dfsg-2 [14.2 kB] 
96% [30 tcl8.6 2,295 B/14.2 kB 16%]                                20.6 MB/s 0s 97% [Working]                                                      20.6 MB/s 0s                                                                                 Get:31 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 thermald amd64 1.5-2ubuntu4 [187 kB] 
97% [31 thermald 2,293 B/187 kB 1%]                                20.6 MB/s 0s 98% [Working]                                                      20.6 MB/s 0s                                                                                 Get:32 http://archive.ubuntu.com/ubuntu xenial/main amd64 tk8.6 amd64 8.6.5-1 [12.2 kB] 
98% [32 tk8.6 2,295 B/12.2 kB 19%]                                 20.6 MB/s 0s 98% [Working]                                                      20.6 MB/s 0s                                                                                 Get:33 http://archive.ubuntu.com/ubuntu xenial/main amd64 xterm amd64 322-1ubuntu1 [607 kB] 
98% [33 xterm 2,293 B/607 kB 0%]                                   20.6 MB/s 0s 99% [Working]                                                      20.6 MB/s 0s                                                                                 Fetched 131 MB in 7s (17.2 MB/s) 
W: --force-yes is deprecated, use one of the options starting with --allow instead. 
E: Failed to fetch https://nvidia.github.io/libnvidia-container/ubuntu16.04/amd64/./libnvidia-container1_1.0.7-1_amd64.deb  gnutls_handshake() failed: Error in the pull function. 
 
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing? 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ezyang commented:
> An alternative for your use case might be to have a local helper function in your test that returns a c10::optional and is either set or not set.

I'm not 100% familiar with the vagaries of C++ here, but it doesn't seem like move elision applies here: https://godbolt.org/z/58o8N9

ezyang committed:
Update on "Make DispatchKeyGuards accept DispatchKey::Undefined"


When working on the vmap prototype I noticed that this was helpful
as it lets me easily initialize a no-op guard, if I need to do it
at constructor time (which I usually do, because the guards don't
have move constructors).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628092](https://our.internmc.facebook.com/intern/diff/D19628092)

[ghstack-poisoned]

ezyang committed:
Update on "Make DispatchKeyGuards accept DispatchKey::Undefined"


When working on the vmap prototype I noticed that this was helpful
as it lets me easily initialize a no-op guard, if I need to do it
at constructor time (which I usually do, because the guards don't
have move constructors).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628092](https://our.internmc.facebook.com/intern/diff/D19628092)

[ghstack-poisoned]
